### PR TITLE
Exploration of different config + expose webpack config

### PR DESCRIPTION
**Pull request by CompuIves**
We currently use `package.json` for our configuration. I think we should move away from using `package.json` and use something like `next.config.js` since it gives us more power than JSON.

An example is the issue for custom Webpack support: https://github.com/zeit/next.js/issues/40.
With a config like this you could do this for example:

```js
// next.config.js

export default {
  webpack: (webpackConfig) => {
    const newConfig = { ...webpackConfig };
    newConfig.module.preloaders.push({ test: /\.js$/, loader: 'eslint-loader' });
    return newConfig;
  },
  cdn: false
}
```

Which is in my opinion better than creating a new file for webpack since this is more centralised. This does however still give users the option to use different files, for the webpack example:

```js
// next.config.js

export default {
  webpack: require('./webpack').default,
  cdn: false
}
```

These are just my thoughts, I found out that I needed a sort of config file in `.js` when starting a PR for this project. Very curious what others think of this 😄.

### CODE REVIEW

**Comment by frol** [comment]
*Disclaimer: I am not a JS developer, so I am not familiar with the traditions in JS world.*

Config files in JS enable great flexibility, but at a cost of greater complexity and potential misuse and errors. In Python (which is my preferred language), it is common to use config files written in Python.

Overall, I like this idea.

File: server/config.js
Comment (frol) [inline server/config.js]: FYI, the above comment is irrelevant for the updated implementation.

**Comment by rauchg** [comment]
This is a pretty great idea I think

**Comment by Florian-R** [comment]
If this one get merged (and I hope it will), should this fall back to the old behaviour, so this is not a breaking change?

**Comment by CompuIves** [comment]
@Florian-R You mean also read `package.json` for CND settings? We could do this, although it does add an extra bit of complexity since then we're reading two files for configuration.

**Comment by frol** [comment]
@Florian-R There existed only one config option: `cdn=true`. I don't think it is so much of a trouble to move one line from `package.json` to `next.config.js`.

**Comment by Florian-R** [comment]
@frol You're right I missed that, don't know why I was sure there was others options. This indeed shouldn't break any code, dismiss my previous comments.

**Comment by CompuIves** [comment]
This PR now also includes https://github.com/zeit/next.js/pull/174

File: server/build/webpack.js
Comment (giladgray) [inline server/build/webpack.js]: why is this async? i don't see any way for it to be async.

Comment (CompuIves) [inline server/build/webpack.js]: Oh right! That need was removed since we now use `require`. Good catch, I'll update the PR.

**Comment by rauchg** [comment]
Important: since this effectively deprecates `package.next` (which I'm 100% ok with), we should probably WARN if we find it

File: server/config.js
Comment (arunoda) [inline server/config.js]: It's a pretty good idea if we could log something, when we are using this user supplied webpack config decorator.
So, I suggest to not to define this.
And when we are using this, we can check for webpack decorator and log it.

File: server/build/webpack.js
Comment (arunoda) [inline server/build/webpack.js]: Do this only if user provides a webpack decorator.
If there's a one, we should log it so the user know this is happening.

File: server/config.js
Comment (arunoda) [inline server/config.js]: We we should also read the `package.json` and check for the next config. If it's there, we should get it. (Only the `cdn` field).
And if we found it, we need log a message as it's deprecated.

**Review by arunoda — CHANGES_REQUESTED** [review]
@CompuIves could you check these suggestion.
After that, we could get ship this PR  :)

File: server/config.js
Comment (arunoda) [inline server/config.js]: @nkzawa @rauchg do we have any specific format to log these messages.

**Comment by jonaswindey** [comment]
Could we have a variable className on https://github.com/zeit/next.js/blob/master/lib/document.js#L11 in this config too?

That would cover https://github.com/zeit/next.js/issues/251

File: server/config.js
Comment (nkzawa) [inline server/config.js]: no, I think

**Comment by CompuIves** [comment]
@arunoda Processed your feedback!
Maybe we could make a helper function in the future to show warnings/notices/errors with the use of [chalk](https://github.com/chalk/chalk) to make them stand out more.

@jonaswindey I think we should create a separate PR for that issue so we can put this PR live sooner.

**Comment by rauchg** [comment]
@jonaswindey @CompuIves the right way to address that is #251

**Review by arunoda — APPROVED** [review]

File: server/build/webpack.js
Comment (arunoda) [inline server/build/webpack.js]: Add a "> " before the message. Like this:

```
console.log('> Using Webpack config function defined in next.config.js.')
```

File: server/config.js
Comment (arunoda) [inline server/config.js]: Change this to: 

console.log("> [warn] You are using package.json for next.js config. Use next.config.js instead.");

> Just FYI: Try not to use "Please" in front of messages. It's sounds like commanding. 
> I tend to use that and one of my UK friend asked to drop that. Messages sounds nice after that :)

**Comment by rauchg** [comment]
Unless, of course, it's tractable to reload everything without a full restart

**Review by arunoda — CHANGES_REQUESTED** [review]
This is beautiful. We just need some few messages changes. I've added comments for them.

You may also need to rebase again since there are some conflicts.

Also we need to add a section to doc about configuring webpack config. Here's a suggestion for the layout:
- Add sub section about Customizing webpack config.
- Tell about the main goal of this.
- Show a simple example (like importing images directly)
- Then, ask to use this with care. (Since user could broke our default config)
- So show safe operations like adding plugins, aliases, loaders etc.

Also try to keep the doc for this minimal and we can have a full detailed section on the Wiki.

File: server/config.js
Comment (CompuIves) [inline server/config.js]: I see, thanks for the tip!

**Comment by CompuIves** [comment]
@arunoda Added docs, let me know what you think!

@rauchg Hmm, this is a valid point. I'll check if there is a way to restart the server when the config file changes.

**Comment by arunoda** [comment]
@rauchg That's a good point, but it seems like it's [hard to do that](http://stackoverflow.com/questions/31472096/watch-webpack-config-js-and-re-run-webpack-command-in-response-to-a-file-change) via webpack itself.

We have to do a hard process restart, may be we could do it using the grunt config easily.

**Comment by arunoda** [comment]
@CompuIves yep. Changes are great. Shall we try @rauchg's suggestion.

**Comment by CompuIves** [comment]
@arunoda Added config watching, it kills the old child process and start a new one when it sees a change in `next.config.js` (only for dev mode).

**Comment by nkzawa** [comment]
I wonder if we should:

- transpile the `next.config.js` (support `import` `export`, `async/await` etc)
- support asynchronous operation in each config

```js
async webpack (webpackConfig, hotReload) {
   const newConfig = { ...webpackConfig }
   newConfig.something = await getSomething()
   return newConfig
 }
```

File: server/build/webpack.js
Comment (nkzawa) [inline server/build/webpack.js]: I think we'd like to pass the `dev` option too.

Comment (CompuIves) [inline server/build/webpack.js]: Right, I changed the second param of `webpackConfig` to `{ hotReload, dev }` so we are more consistent with `createCompiler`.

**Comment by nkzawa** [comment]
Also I wonder how `babel` should be configured.
Babel plugins are set via `babel-loader` on `webpack`, but I believe we wouldn't want users to modify the config object of `webpack` for setting babel plugins.

**Comment by CompuIves** [comment]
@nkzawa We could also add a `babel` option to the config file, which works the same as the `webpack` config. It receives the current babel config as first argument and the options as second argument, what is returned will be used.

**Comment by CompuIves** [comment]
@nkzawa For transpiling `next.config.js` we could use something like [babel-register](https://babeljs.io/docs/usage/babel-register/). Maybe this affects performance a bit, but the config is cached and we read it only once when initializing the application.

File: server/config.js
Comment (CompuIves) [inline server/config.js]: @nkzawa What do you think of this solution?

Comment (nkzawa) [inline server/config.js]: 👍  This is awesome.

But I feel it's kind of complicated that `next.config.js` is transpiled using the default babel settings even if you customized it.
So it might be better not to transpile this file. Node.js supports almost all useful ES6 features now anyway.

Comment (CompuIves) [inline server/config.js]: I agree that this is a bit complicated because of the config, although I do like the idea that we can give the user the option to return a `Promise` instead of a function so they can do asynchronous stuff. I'll add this.

File: README.md
Comment (giladgray) [inline README.md]: rewording proposal:

> You can define custom configuration in a file called `next.config.js` in the project root directory.

File: README.md
Comment (giladgray) [inline README.md]: why do you copy the object first? probably worth adding a code comment

File: README.md
Comment (giladgray) [inline README.md]: sorry for the grammar nags but there should also be a comma _after_ each comma phrase:

> , which is the config used by Next.js[,]

**Comment by giladgray** [comment]
🎉  love how this is coming together!

**Comment by arunoda** [comment]
@nkzawa @CompuIves shall we defer the babel transformation to a new PR (even if someone asked for that)

For Storybook, we don't support it(yet) and there's no problem with that either.

Most importantly, if we defer it we could ship this quickly.

File: README.md
Comment (CompuIves) [inline README.md]: Valid point. I added it as a best practice, because I don't like editing arguments directly.
I do however think this part adds an unnecessary complexity to the documentation, since here it is not necessary to copy first. It'll only raise question marks.

I'll remove it from the documentation.

**Comment by nkzawa** [comment]
btw, It might be better to upgrade wepack to v2 before releasing this PR if we're going to do in the near future.

**Comment by CompuIves** [comment]
@nkzawa You mean so we don't have to make the move to webpack2 a major version bump instead of a minor one?

**Comment by lukeed** [comment]
Love this~! Because exposing `webpack` also lets us **alias React** to something else; eg, Inferno and Preact!

Makes #20 more tangible / within reach! 🙌

**Comment by rauchg** [comment]
Indeed @lukeed :)

**Comment by arunoda** [comment]
@lukeed that's true.
Currently we ship a bundle of all the Next.js related to modules with any app. That includes React as well.

We could configure our app to server React like frameworks, (with some effort) but your app will still load React and React DOM.

So, we should provide a easy way to do this. May be some config in the `next.config.js` for asking not to generate a new bundle.

**Comment by arunoda** [comment]
Guys, this PR is almost ready.
Currently we are working with #310 and it might introduce some changes which affects here. For an example we removed the `hotReload` option.

And it's also nice to wait for https://github.com/zeit/next.js/pull/287 as well. (Where we move into Webpack 2)

**Comment by lukeed** [comment]
> Maybe some config in the `next.config.js` for asking not to generate a new bundle.

@arunoda Or just always build the bundle on `next build` using the merged `default+user` config. At that point, it's the user's responsibility to make sure all the imports are covered (`inferno-compat` / `preact-compat`). Then the pages, components, _and_ bundle are all built & running on the same libraries.

**Comment by developit** [comment]
So this is pretty exciting 🙌

**Comment by arunoda** [comment]
@lukeed That's a pretty good idea. I think we should look at this.
@rauchg what do you think?

**Comment by rauchg** [comment]
Yep. I realize the CDN gets in the way. We'll just use it for the vanilla configuration. That said, we still need to decide the *exact* API for loading another renderer. Because if we have a package like `next-preact` let's say, it could receive the config to decorate it prior to handing it to the user:

```
user(preact(cfg))
```

Then `preact()` could set `cfg.cdn = 'https://cdn.nextjs.org/preact/x-y-z.js` in addition to extending `babel` and `webpack`.

That way we still get the ideal situation where the preact next bundle is shared by all users specially on mobile :)

**Comment by lukeed** [comment]
I feel that if there is still an end-goal or desire to host a `preact` or `inferno` bundle on the CDN, those should just be a pre-built, official bundle alternatives.

That way, there's nothing more a user has to configure beyond `config.cdn = 'inferno'`. Not only does that map to a cdn link, it determines which local file is copied to the `.next` directory as `next.bundle.js`. Since it's local, shared caching doesn't come into play. Ofc, this happens during the `next build` phase.

This removes the need for custom `next-*` adapters. And by exposing the webpack config, users can continue aliasing React to whatever else they please throughout their application builds.

One will only need to be mindful of what versions the alternative bundles are built with.

**Comment by rauchg** [comment]
We don't know that new ones won't emerge, however. The idea of officially maintained `next-preact` and `next-inferno` packages sounds very appealing and is something I will consider. In the meantime, let's continue with the suggestion of turning off the CDN automatically upon override of either `babel` or `webpack`, and let's not forget to document it in the relevant section of the README when it comes time to actualize it.

**Comment by rauchg** [comment]
And by officially maintained I mean keeping the repositories, and then asking @developit and @trueadm for help when necessary 😅

**Comment by developit** [comment]
I do like the middleware idea, where the plugin is just handed the configuration to decorate. That seems the most open-ended and flexible.

**Comment by CompuIves** [comment]
I also like the middleware idea, this way people can support other functionality (like typescript) by writing a middleware for the webpack config.

**Comment by lukeed** [comment]
Status on this? Think it's pretty crucial imo.

Would also like to clarify that I _do_ think a middleware is the way to go. I was just toying with the "pre-built" idea as it came to me. 😇

**Comment by rauchg** [comment]
@lukeed we're merging this ASAP :)

**Comment by rauchg** [comment]
I'd just like to have Webpack 2 support, for which we need this fix to land:
https://github.com/webpack/watchpack/pull/33

cc @nkzawa

**Comment by CompuIves** [comment]
@rauchg You can message me when I can rebase with master + do remaining changes before merge. 😄

**Comment by lukeed** [comment]
@rauchg @arunoda  Sorry to keep coming back 😸 

Would my [suggestion from above](https://github.com/zeit/next.js/pull/222#issuecomment-264375520) rely on integrated Webpack2 first? AFAIK, seems like it could be done now.

If this is still worth considering, I can get an issue open at the very least...

**Comment by arunoda** [comment]
@lukeed You are correct.
Let's open an issue for your request and I like it.

**Comment by aesopwolf** [comment]
FWIW the watchpack fix landed earlier today https://github.com/webpack/watchpack/pull/33

So that shouldn't be holding up Webpack 2 support anymore.

**Comment by rauchg** [comment]
@aesopwolf it's not released unfortunately

**Review by arunoda — APPROVED** [review]

File: server/config.js
Comment (arunoda) [inline server/config.js]: With #414 now we don't have an option called `cdn`. So we don't need this anymore.

File: server/config.js
Comment (arunoda) [inline server/config.js]: Since we are no longer have a `cdn` there's no need to look at the `package.json` or warn the user.

**Review by arunoda — CHANGES_REQUESTED** [review]
Wait until https://github.com/zeit/next.js/pull/414 lands to master.

> EDIT: #414 landed into master

After that, resolve conflicts and implement these suggestions. Then we are good to go with this PR.

**Comment by rauchg** [comment]
✌️️

**Comment by arunoda** [comment]
@lukeed now with https://github.com/zeit/next.js/pull/414 we don't have `next-bundle`. So, it's totally possible to do anything with the webpack config.
So, no need to fire an new issue. What you suggested is implemented :)

**Comment by lukeed** [comment]
@arunoda perfect! 😀

**Comment by CompuIves** [comment]
Rebased master and done the required changes.

I stumbled upon the item of CDN in the FAQ of the README. Do we still need this?

File: README.md
Comment (arunoda) [inline README.md]: We don't have a `cdn` option. May be we need to re-phrase this.

**Review by arunoda — APPROVED** [review]

**Comment by arunoda** [comment]
@CompuIves I added some comments.

**Comment by rauchg** [comment]
I'll remove the cdn notes myself

**Comment by rauchg** [comment]
Thanks @CompuIves

**Comment by arunoda** [comment]
Awesome!

**Comment by CompuIves** [comment]
Nice! Was great to help! :tada:

**Pull request merged**
